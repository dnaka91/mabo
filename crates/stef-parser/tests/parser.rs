use std::{
    fmt::{self, Display},
    fs,
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use stef_parser::Schema;

#[test]
fn parse_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl<'a> Display for Wrapper<'a> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &*value).to_string());
        });
    });
}
