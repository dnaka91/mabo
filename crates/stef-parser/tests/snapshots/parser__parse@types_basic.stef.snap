---
source: crates/stef-parser/tests/parser.rs
description: "struct Sample {\n    f01: bool @1,\n    f02: u8 @2,\n    f03: u16 @3,\n    f04: u32 @4,\n    f05: u64 @5,\n    f06: u128 @6,\n    f07: i8 @7,\n    f08: i16 @8,\n    f09: i32 @9,\n    f10: i64 @10,\n    f11: i128 @11,\n    f12: f32 @12,\n    f13: f64 @13,\n    f14: string @14,\n    f15: &string @15,\n    f16: bytes @16,\n    f17: &bytes @17,\n    f18: box<string> @18,\n    f19: box<bytes> @19,\n    f20: (u32, u32, u32) @20,\n    f21: [u32; 12] @21,\n}"
input_file: crates/stef-parser/tests/inputs/types_basic.stef
---
Schema {
    path: Some(
        "types_basic.stef",
    ),
    source: "struct Sample {\n    f01: bool @1,\n    f02: u8 @2,\n    f03: u16 @3,\n    f04: u32 @4,\n    f05: u64 @5,\n    f06: u128 @6,\n    f07: i8 @7,\n    f08: i16 @8,\n    f09: i32 @9,\n    f10: i64 @10,\n    f11: i128 @11,\n    f12: f32 @12,\n    f13: f64 @13,\n    f14: string @14,\n    f15: &string @15,\n    f16: bytes @16,\n    f17: &bytes @17,\n    f18: box<string> @18,\n    f19: box<bytes> @19,\n    f20: (u32, u32, u32) @20,\n    f21: [u32; 12] @21,\n}\n",
    definitions: [
        Struct(
            Struct {
                comment: Comment(
                    [],
                ),
                attributes: Attributes(
                    [],
                ),
                name: Name {
                    value: "Sample",
                    span: Span {
                        start: 7,
                        end: 13,
                    },
                },
                generics: Generics(
                    [],
                ),
                fields: Named(
                    [
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f01",
                                span: Span {
                                    start: 20,
                                    end: 23,
                                },
                            },
                            ty: Type {
                                value: Bool,
                                span: Span {
                                    start: 25,
                                    end: 29,
                                },
                            },
                            id: Id {
                                value: 1,
                                span: Span {
                                    start: 30,
                                    end: 32,
                                },
                            },
                            span: Span {
                                start: 20,
                                end: 32,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f02",
                                span: Span {
                                    start: 38,
                                    end: 41,
                                },
                            },
                            ty: Type {
                                value: U8,
                                span: Span {
                                    start: 43,
                                    end: 45,
                                },
                            },
                            id: Id {
                                value: 2,
                                span: Span {
                                    start: 46,
                                    end: 48,
                                },
                            },
                            span: Span {
                                start: 38,
                                end: 48,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f03",
                                span: Span {
                                    start: 54,
                                    end: 57,
                                },
                            },
                            ty: Type {
                                value: U16,
                                span: Span {
                                    start: 59,
                                    end: 62,
                                },
                            },
                            id: Id {
                                value: 3,
                                span: Span {
                                    start: 63,
                                    end: 65,
                                },
                            },
                            span: Span {
                                start: 54,
                                end: 65,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f04",
                                span: Span {
                                    start: 71,
                                    end: 74,
                                },
                            },
                            ty: Type {
                                value: U32,
                                span: Span {
                                    start: 76,
                                    end: 79,
                                },
                            },
                            id: Id {
                                value: 4,
                                span: Span {
                                    start: 80,
                                    end: 82,
                                },
                            },
                            span: Span {
                                start: 71,
                                end: 82,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f05",
                                span: Span {
                                    start: 88,
                                    end: 91,
                                },
                            },
                            ty: Type {
                                value: U64,
                                span: Span {
                                    start: 93,
                                    end: 96,
                                },
                            },
                            id: Id {
                                value: 5,
                                span: Span {
                                    start: 97,
                                    end: 99,
                                },
                            },
                            span: Span {
                                start: 88,
                                end: 99,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f06",
                                span: Span {
                                    start: 105,
                                    end: 108,
                                },
                            },
                            ty: Type {
                                value: U128,
                                span: Span {
                                    start: 110,
                                    end: 114,
                                },
                            },
                            id: Id {
                                value: 6,
                                span: Span {
                                    start: 115,
                                    end: 117,
                                },
                            },
                            span: Span {
                                start: 105,
                                end: 117,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f07",
                                span: Span {
                                    start: 123,
                                    end: 126,
                                },
                            },
                            ty: Type {
                                value: I8,
                                span: Span {
                                    start: 128,
                                    end: 130,
                                },
                            },
                            id: Id {
                                value: 7,
                                span: Span {
                                    start: 131,
                                    end: 133,
                                },
                            },
                            span: Span {
                                start: 123,
                                end: 133,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f08",
                                span: Span {
                                    start: 139,
                                    end: 142,
                                },
                            },
                            ty: Type {
                                value: I16,
                                span: Span {
                                    start: 144,
                                    end: 147,
                                },
                            },
                            id: Id {
                                value: 8,
                                span: Span {
                                    start: 148,
                                    end: 150,
                                },
                            },
                            span: Span {
                                start: 139,
                                end: 150,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f09",
                                span: Span {
                                    start: 156,
                                    end: 159,
                                },
                            },
                            ty: Type {
                                value: I32,
                                span: Span {
                                    start: 161,
                                    end: 164,
                                },
                            },
                            id: Id {
                                value: 9,
                                span: Span {
                                    start: 165,
                                    end: 167,
                                },
                            },
                            span: Span {
                                start: 156,
                                end: 167,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f10",
                                span: Span {
                                    start: 173,
                                    end: 176,
                                },
                            },
                            ty: Type {
                                value: I64,
                                span: Span {
                                    start: 178,
                                    end: 181,
                                },
                            },
                            id: Id {
                                value: 10,
                                span: Span {
                                    start: 182,
                                    end: 185,
                                },
                            },
                            span: Span {
                                start: 173,
                                end: 185,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f11",
                                span: Span {
                                    start: 191,
                                    end: 194,
                                },
                            },
                            ty: Type {
                                value: I128,
                                span: Span {
                                    start: 196,
                                    end: 200,
                                },
                            },
                            id: Id {
                                value: 11,
                                span: Span {
                                    start: 201,
                                    end: 204,
                                },
                            },
                            span: Span {
                                start: 191,
                                end: 204,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f12",
                                span: Span {
                                    start: 210,
                                    end: 213,
                                },
                            },
                            ty: Type {
                                value: F32,
                                span: Span {
                                    start: 215,
                                    end: 218,
                                },
                            },
                            id: Id {
                                value: 12,
                                span: Span {
                                    start: 219,
                                    end: 222,
                                },
                            },
                            span: Span {
                                start: 210,
                                end: 222,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f13",
                                span: Span {
                                    start: 228,
                                    end: 231,
                                },
                            },
                            ty: Type {
                                value: F64,
                                span: Span {
                                    start: 233,
                                    end: 236,
                                },
                            },
                            id: Id {
                                value: 13,
                                span: Span {
                                    start: 237,
                                    end: 240,
                                },
                            },
                            span: Span {
                                start: 228,
                                end: 240,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f14",
                                span: Span {
                                    start: 246,
                                    end: 249,
                                },
                            },
                            ty: Type {
                                value: String,
                                span: Span {
                                    start: 251,
                                    end: 257,
                                },
                            },
                            id: Id {
                                value: 14,
                                span: Span {
                                    start: 258,
                                    end: 261,
                                },
                            },
                            span: Span {
                                start: 246,
                                end: 261,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f15",
                                span: Span {
                                    start: 267,
                                    end: 270,
                                },
                            },
                            ty: Type {
                                value: StringRef,
                                span: Span {
                                    start: 272,
                                    end: 279,
                                },
                            },
                            id: Id {
                                value: 15,
                                span: Span {
                                    start: 280,
                                    end: 283,
                                },
                            },
                            span: Span {
                                start: 267,
                                end: 283,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f16",
                                span: Span {
                                    start: 289,
                                    end: 292,
                                },
                            },
                            ty: Type {
                                value: Bytes,
                                span: Span {
                                    start: 294,
                                    end: 299,
                                },
                            },
                            id: Id {
                                value: 16,
                                span: Span {
                                    start: 300,
                                    end: 303,
                                },
                            },
                            span: Span {
                                start: 289,
                                end: 303,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f17",
                                span: Span {
                                    start: 309,
                                    end: 312,
                                },
                            },
                            ty: Type {
                                value: BytesRef,
                                span: Span {
                                    start: 314,
                                    end: 320,
                                },
                            },
                            id: Id {
                                value: 17,
                                span: Span {
                                    start: 321,
                                    end: 324,
                                },
                            },
                            span: Span {
                                start: 309,
                                end: 324,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f18",
                                span: Span {
                                    start: 330,
                                    end: 333,
                                },
                            },
                            ty: Type {
                                value: BoxString,
                                span: Span {
                                    start: 335,
                                    end: 346,
                                },
                            },
                            id: Id {
                                value: 18,
                                span: Span {
                                    start: 347,
                                    end: 350,
                                },
                            },
                            span: Span {
                                start: 330,
                                end: 350,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f19",
                                span: Span {
                                    start: 356,
                                    end: 359,
                                },
                            },
                            ty: Type {
                                value: BoxBytes,
                                span: Span {
                                    start: 361,
                                    end: 371,
                                },
                            },
                            id: Id {
                                value: 19,
                                span: Span {
                                    start: 372,
                                    end: 375,
                                },
                            },
                            span: Span {
                                start: 356,
                                end: 375,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f20",
                                span: Span {
                                    start: 381,
                                    end: 384,
                                },
                            },
                            ty: Type {
                                value: Tuple(
                                    [
                                        Type {
                                            value: U32,
                                            span: Span {
                                                start: 387,
                                                end: 390,
                                            },
                                        },
                                        Type {
                                            value: U32,
                                            span: Span {
                                                start: 392,
                                                end: 395,
                                            },
                                        },
                                        Type {
                                            value: U32,
                                            span: Span {
                                                start: 397,
                                                end: 400,
                                            },
                                        },
                                    ],
                                ),
                                span: Span {
                                    start: 386,
                                    end: 401,
                                },
                            },
                            id: Id {
                                value: 20,
                                span: Span {
                                    start: 402,
                                    end: 405,
                                },
                            },
                            span: Span {
                                start: 381,
                                end: 405,
                            },
                        },
                        NamedField {
                            comment: Comment(
                                [],
                            ),
                            name: Name {
                                value: "f21",
                                span: Span {
                                    start: 411,
                                    end: 414,
                                },
                            },
                            ty: Type {
                                value: Array(
                                    Type {
                                        value: U32,
                                        span: Span {
                                            start: 417,
                                            end: 420,
                                        },
                                    },
                                    12,
                                ),
                                span: Span {
                                    start: 416,
                                    end: 425,
                                },
                            },
                            id: Id {
                                value: 21,
                                span: Span {
                                    start: 426,
                                    end: 429,
                                },
                            },
                            span: Span {
                                start: 411,
                                end: 429,
                            },
                        },
                    ],
                ),
            },
        ),
    ],
}
